include/mbgl/gfx/vertex_attribute.hpp

```diff
@@ -0,0 +1,359 @@
+    std::vector<std::string> readDataDrivenPaintProperties(const Binders& binders, const Evaluated& evaluated) {
+        std::vector<std::string> propertiesAsUniforms;
+        propertiesAsUniforms.reserve(sizeof...(DataDrivenPaintProperty));
+        (
+            [&](const auto& attributeNames) {
+                for (std::size_t attrIndex = 0; attrIndex < attributeNames.size(); ++attrIndex) {
+                    const auto& attributeName = std::string(attributeNames[attrIndex]);
+                    if (auto& binder = binders.template get<DataDrivenPaintProperty>()) {
+                        using Attribute = typename DataDrivenPaintProperty::Attribute;
+                        using Type = typename Attribute::Type; // ::mbgl::gfx::AttributeType<type_, n_>
+                        using InterpType = ZoomInterpolatedAttributeType<Type>;
+                        using Value = typename Type::Value;             // std::array<T, N>
+                        using InterpValue = typename InterpType::Value; // std::array<T, 2*N>
+
+                        const auto vertexCount = binder->getVertexCount();
+                        const auto isConstant = evaluated.template get<DataDrivenPaintProperty>().isConstant();
+                        if (vertexCount > 0 && !isConstant) {
+                            if (auto& attr = getOrAdd("a_" + attributeName)) {
+                                if (const auto& sharedVector = binder->getSharedVertexVector()) {
+                                    const auto rawSize = static_cast<uint32_t>(sharedVector->getRawSize());
+                                    const bool isInterpolated = binder->isInterpolated();
+                                    const auto dataType = isInterpolated ? InterpType::DataType : Type::DataType;
+                                    assert(rawSize ==
+                                           static_cast<uint32_t>(isInterpolated ? sizeof(InterpValue) : sizeof(Value)));
+                                    assert(sharedVector->getRawCount() == vertexCount);
+                                    attr->setSharedRawData(std::move(sharedVector), 0, 0, rawSize, dataType);
+                                } else {
+                                    for (std::size_t i = 0; i < vertexCount; ++i) {
+                                        attr->set(i, binder->getVertexValue(i), attrIndex);
+                                    }
+                                }
+                            }
+                            propertiesAsUniforms.emplace_back(std::string());
+                        } else {
+                            propertiesAsUniforms.emplace_back(attributeName);
+                        }
+                    }
+                }
+            }(DataDrivenPaintProperty::AttributeNames),
+            ...);
+        return propertiesAsUniforms;
+    }
+
+    virtual void copy(const VertexAttributeArray& other) {
```

Comment:
Needs a documentation comment because that includes a description of what happens when a key is already present.